### ISSUE TITLE
Retry support

### ISSUE THREAD

**Issue by kovacs** [easymarketing/sidekiq_workflows#10]
Saw in the notes that retried aren't supported. I'm just getting started with this gem and want to use it for my work and am wondering if you had any insights as to if/when retries would be supported and any context around the challenges or tasks for doing so. If you're not planning, or working on adding support, I'm likely to look at adding it myself assuming I go forward with this gem but wanted to file this issue to gain some insight into the things mentioned above, both for myself and for others that come across this and likely have the same need.

**Comment by foxtacles** [comment]
Hi! I think it should now be possible to add support for retries since Sidekiq Pro introduced the `death` callback:

https://github.com/mperham/sidekiq/blob/master/Pro-Changes.md#403

When we implemented this gem there was no such callback, and it was basically impossible to know when a batch had ultimately failed: the `complete` callback would run as soon as each job ran once (successful or not), and the `successful` callback would only run if all jobs would complete without an error. The `death` callback now fills this gap, and I think this gem may be modified to now use the `success` and `death` callbacks, respectively, instead of `complete` .

The main issue I see is that the `death` state is not propagated to parent batches:
https://github.com/mperham/sidekiq/issues/4774#issuecomment-756320927

So we cannot use this mechanism in the "root" workflow batch anymore:
https://github.com/easymarketing/sidekiq_workflows/blob/master/lib/sidekiq_workflows/worker.rb#L53

because `complete` would no longer be appropriate if we allow retries, `success` would only fire if the entire workflow succeeds, and `death` would not fire because it's not propagated. Basically the same issue as @Draiken is describing. We'd have to work around it by firing this callback in the sub batches whenever it occurs, but also make sure it only fires exactly once per workflow.

To summarize, I think it would be great if we can introduce support for retries but retaining the current semantics may be difficult.

**Comment by foxtacles** [comment]
Addendum: it seems this issue has actually been fixed in the very latest version of Pro:

https://github.com/mperham/sidekiq/blob/master/Pro-Changes.md#521

So I think we can go forward with implementing retries, although we'd have to bump the minimum Sidekiq version requirement to Sidekiq 6 for this, but I'd be okay with that.

**Comment by kovacs** [comment]
I ended up not using this gem and have written my own logic using state machine. It's likely my own deficiency but I wasn't able to get things working fairly easily so I decided to just move forward for now and revisit later. I will need a better workflow tool at some point in the future but I may end up using another solution entirely that's not in ruby. If I have time I'll go and look at this gem again to see where my failure to understand occurred. 

At the moment I was planning on dealing with failure conditions of my batches by looking at the failure count on status object in the complete callback.

**Comment by foxtacles** [comment]
Retries have been implemented in 1.0

**Comment by briankangdev** [comment]
Hi! I have a question regarding the callbacks.
Whne one of the jobs dies in a batch, parent batch's on_complete callback will be executed?

I'm wondering if it will call both callbacks "on_complete" and "on_death".
